File: multiplication/karatsuba.cpp
#include <iostream>
#include<string>
#include <algorithm> 
#include <cmath>
using namespace std;

int karatsuba(int x, int y){
	if ((to_string(x).length()==1) || (to_string(y).length()==1)) {
		return x*y;
  }
	else{
		int n = max(to_string(x).length(),to_string(y).length());
		int z = n / 2;
		int a = x / (int)pow(10,z);
		int b = x % (int)pow(10,z);
		int c = y / (int)pow(10,z);
		int d = y % (int)pow(10,z);
		int ac = karatsuba(a,c);
		int bd = karatsuba(b,d);
		int adbc = karatsuba(a+b,c+d) - ac - bd;
		return ac * pow(10,2*z) + (adbc * pow(10,z)) + bd;
}
}

int main() {
  cout << karatsuba (923,1723);
}

File: multiplication/karatsuba.py
def karatsuba(x,y):
	if len(str(x)) or len(str(y)) :
		return x*y
	else:
		n = max(len(str(x)),len(str(y)))
		z = n / 2
		a = x / 10**(z)
		b = x % 10**(z)
		c = y / 10**(z)
		d = y % 10**(z)
		ac = karatsuba(a,c)
		bd = karatsuba(b,d)
		adbc = karatsuba(a+b,c+d) - ac - bd
		return ac * 10**(2*z) + (adbc * 10**z) + bd

karatsuba (923,1723)